src/adaptation/icap/ResponseParser.cc

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "adaptation/icap/ResponseParser.h"
+#include "parser/Tokenizer.h"
+
+const SBuf Adaptation::Icap::ResponseParser::Icap1magic("ICAP/1.0");
+
+int
+Adaptation::Icap::ResponseParser::parseResponseFirstLine()
+{
+    Tokenizer tok(buf_);
+
+    if (msgProtocol_.protocol != AnyP::PROTO_NONE) {
+        debugs(74, 6, "continue incremental parse for " << msgProtocol_);
+        debugs(74, DBG_DATA, "parse remaining buf={length=" << tok.remaining().length() << ", data='" << tok.remaining() << "'}");
+        // we already found the magic, but not the full line. keep going.
+        return parseResponseStatusAndReason(tok);
+
+    } else if (tok.skip(Icap1magic) && tok.skipOne(DelimiterCharacters())) {
+        debugs(74, 6, "found prefix magic " << Icap1magic);
+        msgProtocol_.protocol = AnyP::PROTO_ICAP;
+        msgProtocol_.major = 1;
+        msgProtocol_.minor = 0;
+
+        // ICAP/1.0 Response status-line parse
+        debugs(74, DBG_DATA, "parse remaining buf={length=" << tok.remaining().length() << ", data='" << tok.remaining() << "'}");
+        buf_ = tok.remaining(); // resume checkpoint
+        return parseResponseStatusAndReason(tok);
+
+    } else if (buf_.length() < Icap1magic.length() && Icap1magic.startsWith(buf_)) {
+        debugs(74, 7, Raw("valid ICAP/1 prefix", buf_.rawContent(), buf_.length()));
+        return 0;
+    }
+
+    // else, protocol error
+    assert(tok.atEnd());
```

Comment:
The overall logic here looks very generic, not specific to ICAP. Please avoid duplicating this code (if it is duplicated) or creating a protocol-specific code instead of generic one.

Reply:
The pattern is very similar to HTTP status-line parsing. But only the first if-statement is actually the same.